### PR TITLE
Fix ARM64 MacOS varargs

### PR DESCRIPTION
**Pull request by keynmol**
This is very rough, just to demonstrate the issue and the solution.

Want to gather feedback on the following:

1. Is it time to refactor and extract common parts from x86 and arm64 versions?
2. Should we handle all platforms in this PR?
3. Is memory padding from the swift code necessary? https://github.com/apple/swift/blob/main/stdlib/public/core/VarArgs.swift#L604-L631
    - I couldn't quite get through the `ifdef`s in swift code, just did this and it worked

TL;DR: varargs are broken on MacOS ARM64

Reason: on most other platforms is a struct (like this one https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#definition-of-va-list) but on this particular one (M1) it's a direct pointer to arguments serialised in memory.

### To reproduce the problem

Use any function with variadic arguments on Apple M1 (SN 0.4.1). I've put this example in a sandbox:

```scala
import scala.scalanative.libc.stdio
import scala.scalanative.unsafe._

object Test {
  def main(args: Array[String]): Unit = {
    Zone { implicit z =>
      val myStr = c"2567 1234"
      val struct = stackalloc[CStruct2[Int, Int]]
      stdio.sscanf(myStr, c"%d %d", struct.at1, struct.at2)

      stdio.printf(c"Original string: %s, parsed.. First: %d, second: %d", myStr, struct._1, struct._2)
    }
  }
}
```

Without the fix in this PR it will segfault upon enountering `sscanf`

With the fix, it will finish successfully:

```
[info] Linking (410 ms)
[info] Checking intermediate code (31 ms)
[info] Dumping intermediate code (linked) (143 ms)
[info] Discovered 638 classes and 4313 methods
[info] Optimizing (debug mode) (370 ms)
[info] Checking intermediate code (20 ms)
[info] Dumping intermediate code (optimized) (159 ms)
[info] Dumping intermediate code (lowered) (272 ms)
[info] Generating intermediate code (537 ms)
[info] Produced 8 files
[info] Compiling to native code (942 ms)
[info] Linking native code (immix gc, none lto) (81 ms)
[info] Total (2713 ms)
Original string: 2567 1234, parsed.. First: 2567, second: 1234[success] Total time: 5 s, completed 7 Dec 2021, 20:32:08
```

### CODE REVIEW

File: nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala
Comment (keynmol) [inline nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala:128]: Copied from the x86_64 unix version

File: nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala
Comment (WojciechMazur) [inline nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala:13]: Linktime resolved values/conditions were designed to dead code eliminate untaken branches at link time, and the main goal was to get rid of usages of native functions that are not available in Windows/Unix. MacOS does not define any os-specific functions that would not be available in other UNIX based operating systems. 
We want to limit the amount of link-time resolved values to a minimum, and use runtime resolved values whenever possible. 
Usage of this method can be replaced with `scala.scalanative.runtime.Platform.isMac`

File: nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala
Comment (WojciechMazur) [inline nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala:21]: Now, after merging https://github.com/scala-native/scala-native/pull/2452 we can detect CPU architecture at runtime in a convenient way in the same way as on the JVM. As long as we don't use any Arm64 specific functions we should use runtime resolved value instead. 
I would suggest moving this method to `scala.scalanative.runtime.PlatformExt` (`Platform` is extern module and it can only contain extern methods, so we can define runtime resolved `object PlatformExt`) where we would use the same logic as on the JVM to check if we are using Arm64 based machine.

File: nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala
Comment (WojciechMazur) [inline nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala:132]: Since the only change is the absence of the header, let's keep it simple with only a single implementation for now, in case if in the future we would locate more inconsistencies we might refactor it to make addressing them easier. For now, I believe we should just rename previous method to `toCVarArgList_Unix` and use the runtime condition to check if the header is used.

**Review by WojciechMazur — CHANGES_REQUESTED** [review]

**Comment by WojciechMazur** [comment]
Thank you for your contribution! I've left some comments in the review for you

File: nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala
Comment (keynmol) [reply nativelib/src/main/scala/scala/scalanative/meta/LinktimeInfo.scala:13]: Gotcha, thanks

**Comment by keynmol** [comment]
Oops, still not right, and local compilation is acting weirdly for me.
Give me a bit more time.

File: nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala
Comment (WojciechMazur) [inline nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala:7]: Scala Native does not define `java.util.Locale`, but it can be provided via external libraries. For usage inside Scala Native us just `.toLowerCase` it should be compliant with usage of `Locale.ROOT`

File: nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala
Comment (ekrich) [inline nativelib/src/main/scala/scala/scalanative/unsafe/CVarArgList.scala:132]: Can we leave change `isMac()` to `isMac`?

**Review by WojciechMazur — APPROVED** [review]

**Comment by keynmol** [comment]
@WojciechMazur Please don't merge this yet (unless you have a way to test it), as I need to re-test it on my machine which last night refused to cooperate on this PR.

**Comment by WojciechMazur** [comment]
> @WojciechMazur Please don't merge this yet (unless you have a way to test it), as I need to re-test it on my machine which last night refused to cooperate on this PR.

Ok, acknowledged. Thank you for notifying me. We definitely need to find a way to tests ARM in the CI

File: nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala
Comment (keynmol) [inline nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala:9]: I thought I'd confirm it with C:

```
❯ bat test.c
───────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: test.c
───────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ #include <sys/utsname.h>
   2   │  #include <stdio.h>
   3   │  int main(int argc, char *argv[]) {
   4   │     struct utsname buffer;
   5   │     if (uname(&buffer) >= 0) {
   6   │       printf("%s", buffer.machine);
   7   │     }
   8   │
   9   │ }
───────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

❯ clang --version
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

❯ clang test.c

❯ ./a.out
arm64%
```

Comment (WojciechMazur) [reply nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala:9]: What is value of os.arch system property on the JVM? Maybe there is more JVM compliance changes that I initially thought

Comment (keynmol) [reply nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala:9]: Hehe

```scala
❯ scala-cli repl
scala> System.getProperty("os.arch")
val res0: String = aarch64

```

Comment (WojciechMazur) [reply nativelib/src/main/scala/scala/scalanative/runtime/PlatformExt.scala:9]: So it looks like clang and all other decent tools are depending on the `uname`, and we do the same. I've not managed to fully find out how does JDK set up `os.arch` on Unix. On Windows, there are like 4 variants using C precompiler definition, but for Unix, it looks more like manually set values in the build. Guess we need to stick to the `uname` for now.

**Pull request merged**
